### PR TITLE
#18 Implement LCOM5 metric

### PR DESCRIPTION
**Pull request by sergey-karazhenets**

### CODE REVIEW

**Comment by 0crat** [comment]
@yegor256 please, pay attention to this pull request

**Comment by 0crat** [comment]
Job `gh:yegor256/jpeek#52` is in scope.

File: src/test/java/org/jpeek/metrics/cohesion/LCOM5Test.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+/**
+ * Test case for {@link LCOM5}.
+ *
+ * @author Sergey Karazhenets (sergeykarazhenets@gmail.com)
+ * @version $Id$
+ * @since 0.13
+ * @checkstyle AbbreviationAsWordInNameCheck (10 lines)
+ * @checkstyle JavadocMethodCheck (500 lines)
+ * @checkstyle LineLengthCheck (500 lines)
+ * @checkstyle EmptyLinesCheck (500 lines)
```
Comment (yegor256) [inline src/test/java/org/jpeek/metrics/cohesion/LCOM5Test.java]: @sergey-karazhenets please, remove this, it's important to keep lines together: http://www.yegor256.com/2014/11/03/empty-line-code-smell.html

File: src/test/java/org/jpeek/metrics/cohesion/LCOM5Test.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+/**
+ * Test case for {@link LCOM5}.
+ *
+ * @author Sergey Karazhenets (sergeykarazhenets@gmail.com)
+ * @version $Id$
+ * @since 0.13
+ * @checkstyle AbbreviationAsWordInNameCheck (10 lines)
+ * @checkstyle JavadocMethodCheck (500 lines)
+ * @checkstyle LineLengthCheck (500 lines)
```
Comment (yegor256) [inline src/test/java/org/jpeek/metrics/cohesion/LCOM5Test.java]: @sergey-karazhenets please, put this in front of the line you _need_ to make longer than 80 chars

File: src/test/java/org/jpeek/metrics/cohesion/LCOM5Test.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+ * @checkstyle EmptyLinesCheck (500 lines)
+ * @checkstyle MethodBodyCommentsCheck (500 lines)
```
Comment (yegor256) [inline src/test/java/org/jpeek/metrics/cohesion/LCOM5Test.java]: @sergey-karazhenets it's a bad practice to comment the code. If you need to explain how it works, use JavaDoc blocks or just break it down into smaller methods

File: src/main/java/org/jpeek/metrics/cohesion/LCOM5.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        ).xembly();
+    }
+
+    /**
+     * Calculates LCOM5 metric for a single Java class.
+     *
+     * @param ctc The .class file.
+     * @return Metrics.
+     * @checkstyle CyclomaticComplexityCheck (10 lines)
+     * @checkstyle ReturnCountCheck (50 lines)
```
Comment (yegor256) [inline src/main/java/org/jpeek/metrics/cohesion/LCOM5.java]: @sergey-karazhenets many `return` statements per method is a bad idea, don't do it: http://www.yegor256.com/2015/08/18/multiple-return-statements-in-oop.html

File: src/main/java/org/jpeek/metrics/cohesion/LCOM5.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+     * @checkstyle AnonInnerLengthCheck (50 lines)
+     * @checkstyle ParameterNumberCheck (50 lines)
+     * @checkstyle MethodBodyCommentsCheck (50 lines)
```
Comment (yegor256) [inline src/main/java/org/jpeek/metrics/cohesion/LCOM5.java]: @sergey-karazhenets in-method-body comments is a bad practice in general. Instead, either use Javadoc blocks or introduce new methods

File: src/main/java/org/jpeek/metrics/cohesion/LCOM5.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+     */
+    @SuppressWarnings({"PMD.UseVarargs", "PMD.UselessParentheses",
```
Comment (yegor256) [inline src/main/java/org/jpeek/metrics/cohesion/LCOM5.java]: @sergey-karazhenets why `"PMD.UselessParentheses"`? Can't you simply get rid of parentheses?

File: src/main/java/org/jpeek/metrics/cohesion/LCOM5.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        "PMD.UseObjectForClearerAPI", "PMD.NPathComplexity",
+        "PMD.CyclomaticComplexity", "PMD.StdCyclomaticComplexity",
+        "PMD.ModifiedCyclomaticComplexity", "PMD.OnlyOneReturn"})
```
Comment (yegor256) [inline src/main/java/org/jpeek/metrics/cohesion/LCOM5.java]: @sergey-karazhenets get rid of `"PMD.OnlyOneReturn"` here please

**Comment by yegor256** [comment]
@sergey-karazhenets looks cool, but there are a few comments above

**Comment by yegor256** [comment]
@sergey-karazhenets what's up with this pull request?

**Comment by sergey-karazhenets** [comment]
@yegor256 Please sorry for delay, I will certainly make fixes, don't have much time to do that now

**Comment by yegor256** [comment]
@sergey-karazhenets thanks, looking forward to them!

**Comment by sergey-karazhenets** [comment]
@yegor256 please review changes

**Comment by yegor256** [comment]
@sergey-karazhenets Jesus Christ... are you sure we need all these `asm/*` clases? They look very strange to me.

**Comment by sergey-karazhenets** [comment]
I think yes, but let's look together. `Jp*` prefixed classes provide wrap for ASM visitors. They avoid creation of long anonymous classes in each metric and provide possibility to subscribe on byte code traversing events via functions and procedures provided through constructor. We don't need explicitly override methods, just need provide a function which will be called on expecting event. Each function may be decorated to change original behaviour. For example we don't want to consider generated by compiler fields and we write in each metric something like 
```
if (!name.contains("$")) { 
   // do something 
}
```
 Why don't replace it on one function and use function in all metrics?! My changes can reduce complexity of `cohesion` method for each metric and provide possibility to place in one function some common logic between metrics.
Traversing of byte code for `LCOM5` was looked like:
```java
        final ClassReader reader;
        try {
            reader = new ClassReader(ctc.toBytecode());
        } catch (final IOException | CannotCompileException ex) {
            throw new IllegalStateException(ex);
        }
        final List<String> fields = new LinkedList<>();
        final List<Collection<String>> methods = new LinkedList<>();
        reader.accept(
            new ClassVisitor(Opcodes.ASM6) {
                @Override
                public FieldVisitor visitField(
                    final int access, final String name, final String desc,
                    final String signature, final Object value
                ) {
                    if (!name.contains("$")) {
                        fields.add(name);
                    }
                    return null;
                }
                @Override
                public MethodVisitor visitMethod(
                    final int access, final String name, final String desc,
                    final String signature, final String[] exceptions
                ) {
                    if ("<init>".equals(name) || name.contains("$")) {
                        return null;
                    }
                    final Set<String> mfields = new HashSet<>(0);
                    methods.add(mfields);
                    return new MethodVisitor(Opcodes.ASM6) {
                        @Override
                        public void visitFieldInsn(
                            final int opcode, final String owner,
                            final String name, final String desc
                        ) {
                            final String ownerclass = Type
                                .getObjectType(owner).getClassName();
                            if (!name.contains("$")
                                && ownerclass.equals(ctc.getName())) {
                                mfields.add(name);
                            }
                        }
                    };
                }
            },
            0
        );
```
With introduced classes it looks like:
```java
        final List<String> fields = new LinkedList<>();
        final List<Collection<String>> methods = new LinkedList<>();
        new JpClassReader(
            new CtClassByteCode(ctc),
            new JpClassVisitor(
                new OnNotGenField(
                    new WithoutFieldVisitor(
                        (access, name, desc, signature, value)
                            -> fields.add(name)
                    )
                ),
                new OnNotGenMethod(
                    new OnNotCtor(
                        (access, name, desc, signature, exceptions) -> {
                            final Set<String> flds = new HashSet<>(0);
                            methods.add(flds);
                            return new JpMethodVisitor(
                                new OnNotGenFieldInsn(
                                    new OnFieldInsnOwnedByClass(
                                        ctc.getName(),
                                        (opcode, owner, fname, fdesc)
                                            -> flds.add(fname)
                                    )
                                )
                            );
                        }
                    )
                )
        )).value();
```
Which one variant is better?
Second realizes OOP way against procedural in first variant.
In a second variant to filter generated by compiler fields/methods/field instructions inside a method, constructors I use `OnNotGenField`, `OnNotGenMethod`, `OnNotGenFieldInsn`, `OnNotCtor` functions. To react only on all not generated methods and constructors you can compose functions:
```java
new OnNotGenMethod(
    new OnNotCtor(
       (access, name, desc, signature, exceptions) -> /*  do something with method */ 
   )
)
```

Also Javassist provides poorer API to traverse byte code than ASM. Javassist useful when you need to collect info just about fields, methods. When you need collect information about code inside some method in this case ASM more preferable. And let use ASM for all metrics to unify approach of metrics calculation.

**Comment by yegor256** [comment]
@sergey-karazhenets you're right, we need improvements, but the way you approach is too abstract, too generic. See what I did here: https://github.com/yegor256/jpeek/commit/c40269fa4ae77dee7fdfecdf9f16b1a77b057b7e I just took a piece of functionality and placed it into a new class. Just one class, which does specifically one thing -- retrieves methods from a class. Looks like we need a few more "supplementary" classes like that. Pay attention, we don't need blocks, which we will use to construct that classes. We just need classes, for one particular purpose.

**Comment by 0crat** [comment]
Job `gh:yegor256/jpeek#52` assigned to [@janczer](http://www.0crat.com/u/janczer}). The budget is [fixed](http://datum.zerocracy.com/pages/policy.html#4) and it is 30 minutes. Please, read [§4](http://datum.zerocracy.com/pages/policy.html#4), [§8](http://datum.zerocracy.com/pages/policy.html#8), and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by janczer** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/yegor256/jpeek/issues/52#issuecomment-357239046))

@janczer [@janczer](http://www.0crat.com/u/janczer) resigned from [#52](https://github.com/yegor256/jpeek/issues/52), please stop working

**Comment by 0crat** [comment]
The job [#52](https://github.com/yegor256/jpeek/issues/52) assigned to vmotsak. The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by yegor256** [comment]
@eugene-nikolaev this is not valid anymore, since our mechanism of metrics calculation is totally different now. Sorry.

**Comment by eugene-nikolaev** [comment]
I guess that was actually for @sergey-karazhenets

**Comment by 0crat** [comment]
The job [#52](https://github.com/yegor256/jpeek/issues/52) is now out of scope

**Comment by 0crat** [comment]
Order was successfully finished: +30 points just awarded to [@vmotsak](http://www.0crat.com/u/vmotsak), total is [+30](http://www.0crat.com/u/vmotsak)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://datum.zerocracy.com/pages/policy.html#28): +15 points just awarded to [@yegor256](http://www.0crat.com/u/yegor256), total is [+7255](http://www.0crat.com/u/yegor256)

**Pull request closed without merging**